### PR TITLE
chore: updated reanimated to v3

### PR DESCRIPTION
**Pull request by gorhom**
## Motivation

Part of the requirements for #1125 , i have updated reanimated to v3 and react native to 71.

### Known Issues

```
 ERROR  ReanimatedError: Cannot read property '__remoteFunction' of undefined, js engine: reanimated
```

this issue is triggered when calling `Keyboard.dismiss` on JS thread, seems related to https://github.com/software-mansion/react-native-reanimated/issues/4201.

### CODE REVIEW

**Comment by gorhom** [comment]
LGTM 🗿

**Comment by YaoHuiJi** [comment]
hi, @gorhom , I've tried 5.0.0-alpha.0 , but the issue `ERROR  ReanimatedError: Cannot read property '__remoteFunction' of undefined, js engine: reanimated` is still here. I have cleaned the cache and rebuilt the xcode project, but nothing works.

**Comment by gorhom** [comment]
@YaoHuiJi it is related to https://github.com/software-mansion/react-native-reanimated/issues/4201, it only occurs in development

**Comment by YaoHuiJi** [comment]
hi @gorhom ,  I tried to run it in release mode, but the error still seems to be there, here is the xcode log:

```
libc++abi: terminating with uncaught exception of type facebook::jsi::JSError: Cannot read property '__remoteFunction' of undefined

TypeError: Cannot read property '__remoteFunction' of undefined
    at runOnJS (JavaScript:1:83)
    at handleOnEnd (JavaScript:1:1308)
    at anonymous (JavaScript:1:215)
    at apply (native)
    at runWorklet (JavaScript:1:207)
    at anonymous (JavaScript:1:886)
    at handleAndFlushImmediates (JavaScript:1:143)
dyld4 config: DYLD_LIBRARY_PATH=/usr/lib/system/introspection DYLD_INSERT_LIBRARIES=/Developer/usr/lib/libBacktraceRecording.dylib:/Developer/usr/lib/libMainThreadChecker.dylib:/Developer/Library/PrivateFrameworks/DTDDISupport.framework/libViewDebuggerSupport.dylib:/usr/lib/libMTLCapture.dylib
terminating with uncaught exception of type facebook::jsi::JSError: Cannot read property '__remoteFunction' of undefined

TypeError: Cannot read property '__remoteFunction' of undefined
    at runOnJS (JavaScript:1:83)
    at handleOnEnd (JavaScript:1:1308)
    at anonymous (JavaScript:1:215)
    at apply (native)
    at runWorklet (JavaScript:1:207)
    at anonymous (JavaScript:1:886)
    at handleAndFlushImmediates (JavaScript:1:143)
```

In my app, when the keyboard is open( a BottomSheetTextInput in bottom sheet modal has the focus), then If you drag the bottom sheet modal to resize it to another snap point, this error will occur.

**Comment by malikzype** [comment]
@gorhom We face a similar issue on Android and it usually occurs because of BackdropComponent with reanimated v3.0.2. Any plans of rolling v5 out soon?

**Comment by AfrazHussain** [comment]
I didn't get enough time to dig deep into this, but I can confirm that there seems to be issues with the Backdrop, BottomSheetFlatList Components only while using fabric. After I turn fabric off, everything seems to be okay. 
While using fabric and BottomSheetFlatList, I get the error `TypeError: Cannot read property 'stateNode' of undefined`, which was also reported (not by me) in this reanimated issue https://github.com/software-mansion/react-native-reanimated/issues/3534. If I use RN FlatList, then there seems to be no issues.

File: example/app/package.json [hunk 1]
```diff
@@ -18,22 +18,22 @@
...
-    "react": "17.0.2",
-    "react-native": "0.68.1",
-    "react-native-gesture-handler": "~2.8.0",
+    "react": "18.2.0",
+    "react-native": "0.71.3",
+    "react-native-gesture-handler": "~2.9.0",
     "react-native-maps": "^0.30.1",
     "react-native-pager-view": "^5.4.24",
-    "react-native-reanimated": "~2.12.0",
+    "react-native-reanimated": "~2.14.4",
```
Comment (DalYoon) [inline example/app/package.json:26]: Hi 👋, @gorhom  I have a question
Isn't this PR for updating reanimated v3?
I'm wondering why the changed version is v2.14.4 of it.

**Comment by levipro** [comment]
@gorhom I have resolved the error ` ERROR  ReanimatedError: Cannot read property '__remoteFunction' of undefined, js engine: reanimated` by defining the callback for `runOnJS` with an arrow function:

```diff
diff --git a/node_modules/@gorhom/bottom-sheet/src/hooks/useGestureEventsHandlersDefault.tsx b/node_modules/@gorhom/bottom-sheet/src/hooks/useGestureEventsHandlersDefault.tsx
index b44bfa5..0b029d3 100644
--- a/node_modules/@gorhom/bottom-sheet/src/hooks/useGestureEventsHandlersDefault.tsx
+++ b/node_modules/@gorhom/bottom-sheet/src/hooks/useGestureEventsHandlersDefault.tsx
@@ -44,6 +44,10 @@ export const useGestureEventsHandlersDefault: GestureEventsHandlersHookType =
     } = useBottomSheetInternal();
     //#endregion
 
+    const dismissKeyboard = () => {
+      Keyboard.dismiss();
+    };
+
     //#region gesture methods
     const handleOnStart: GestureEventHandlerCallbackType<GestureEventContextType> =
       useWorkletCallback(
@@ -296,7 +300,7 @@ export const useGestureEventsHandlersDefault: GestureEventsHandlersHookType =
                 absoluteY > WINDOW_HEIGHT - animatedKeyboardHeight.value
               )
             ) {
-              runOnJS(Keyboard.dismiss)();
+              runOnJS(dismissKeyboard)();
             }
           }
 

```

**Comment by YaoHuiJi** [comment]
@levipro thank you! it works!

**Comment by fgagneten** [comment]
> I didn't get enough time to dig deep into this, but I can confirm that there seems to be issues with the Backdrop, BottomSheetFlatList Components only while using fabric. After I turn fabric off, everything seems to be okay. While using fabric and BottomSheetFlatList, I get the error `TypeError: Cannot read property 'stateNode' of undefined`, which was also reported (not by me) in this reanimated issue [software-mansion/react-native-reanimated#3534](https://github.com/software-mansion/react-native-reanimated/issues/3534). If I use RN FlatList, then there seems to be no issues.

Same issue using BottomSheetVirtualizedList + Fabric. Using just VirtualizedList from RN fixes the error, but the BottomSheet does not work as expected

**Comment by fgagneten** [comment]
@gorhom I also tested with the last 5.0.0-alpha0 version and same error

**Comment by francozini2** [comment]
> > I didn't get enough time to dig deep into this, but I can confirm that there seems to be issues with the Backdrop, BottomSheetFlatList Components only while using fabric. After I turn fabric off, everything seems to be okay. While using fabric and BottomSheetFlatList, I get the error `TypeError: Cannot read property 'stateNode' of undefined`, which was also reported (not by me) in this reanimated issue [software-mansion/react-native-reanimated#3534](https://github.com/software-mansion/react-native-reanimated/issues/3534). If I use RN FlatList, then there seems to be no issues.
> 
> Same issue using BottomSheetVirtualizedList + Fabric. Using just VirtualizedList from RN fixes the error, but the BottomSheet does not work as expected

+1 Started having the same issue after enabling the new arch

File: example/app/package.json (hunk 1 above)
Comment (MiracleUFO) [reply example/app/package.json:26]: Yes, why is this using reanimated 2.14.4?

**Pull request merged**